src/server.c

```diff
@@ -3943,6 +4436,18 @@ void populateCommandLegacyRangeSpec(struct redisCommand *c) {
     c->legacy_range_key_spec.fk.range.limit = 0;
 }
 
+void commandAddSubcommand(struct redisCommand *parent, struct redisCommand *subcommand) {
+    if (!parent->subcommands_dict)
+        parent->subcommands_dict = dictCreate(&commandTableDictType);
+
+    subcommand->parent = parent; /* Assign the parent command */
+    sds fullname = getFullCommandName(subcommand);
```

Comment:
can we just set the `subcommand->name = fullname`? (or add a fullname field....)
it seems to me now that the subcommand name are generally unusable, we must get the fullname to use it.
so i suppose if we can just set the name = fullname, when we need it, we can just use it...

like here, it looks like the subcommand won't get the full name (unfamiliar with modules)
https://github.com/redis/redis/blob/56a802057e90ea482702dbd5191bc397caf0f596/src/module.c#L10198-L10204

Reply:
interesting..
in theory changing that (past 7.0) could be a breaking change.
but currently (before 7.0 is released), modules can't register sub-commands.

@enjoy-binbin did you conduct some search to see who's still using the simple `cmd->name` (on something that could be a sub-command) to find that one?
if that's the only one, i tend to agree we wanna make `name` the full name by default and drop all the complication of `getFullCommandName`getFullCommandName

Reply:
> did you conduct some search to see who's still using the simple cmd->name (on something that could be a sub-command) to find that one?

actually no, i happen to see it (reviewing / studying some module codes.)

i also tend to use the full name by default, so that we don't have to deal with `getFullCommandName`.
i do some search, here are what i found: 


1. latencyAllCommandsFillCDF, #10103

2. afterErrorReply (we need to take care of it, it could be a subcommand i think):
```
const char *cmdname = c->lastcmd ? c->lastcmd->name : "<unknown>";
serverLog(LL_WARNING,"== CRITICAL == This %s is sending an error "
                     "to its %s: '%.*s' after processing the command "
                     "'%s'", from, to, (int)len, s, cmdname);
```

3. ACL CAT? maybe we need to also print the subcommand in future? (for now, it only works on parentcommand)

4. COMMAND LIST? same as above, only print the parent command, but others one, like command docs config|get, can work on subcommand

Reply:
@guybe7 considering sub-commands are now first class citizens, i suppose we should list them in ACL CAT and COMMAND LIST, WDYT?

and IIRC from @enjoy-binbin findings, we might be better off just deleting `getFullCommandName` and put the full command name in `cmd->name`, which will also fix:
* RM_GetCurrentCommandName module API
* the `CRITICAL` log warning in `afterErrorReply
* replies in `COMMAND LIST` and `ACL CAT` once we change them to include sub-commands.

Reply:
@enjoy-binbin Guy didn't respond yet, but i'm paranoid about forgetting this (discussion in a closed PR), and i think we must do that in 7.0 for completeness. wanna make a PR and we can continue the discussion there if needed?

Reply:
sure, i can handle `getFullCommandName` first (i will keep in mind :)

Reply:
@oranagra @enjoy-binbin I agree. we should always store the fullname, and eliminate `getFullCommandName`.
we should rename `redisCommand.name` to something like `tmp_cmdname` so that the compiler will make sure we didn't miss anything. after that, we can call it `name` again. 9another option is to change it to `fullname` permanently in order to cause conflicts in open PRs such as https://github.com/redis/redis/pull/10103)

Reply:
i think that the command name is *always* the full name (also in the ubcommand_dict)
to clarify, the word "rewrite" from `CONFIG REWRITE` should *never* appear anywhere by itself, it should always be saved as "config|rewrite"

Reply:
suggestion: we need to make sure that all callers to `commandAddSubcommand` already pass the subcommand with the fullname

Reply:
@enjoy-binbin maybe while you're working on it, you can also handle #10124 and see if we missed anything.